firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -33,6 +35,115 @@
   /** The event listener for the query that raises events asynchronously. */
   private final ExecutorEventListener<ViewSnapshot> asyncEventListener;
 
+  public static class StopListenerSupportFragment extends android.support.v4.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      for (Runnable callback : callbacks) {
+        if (callback != null) {
+          callback.run();
+        }
+      }
+    }
+
+    void addCallback(Runnable callback) {
+      callbacks.add(callback);
+    }
+
+    private List<Runnable> callbacks = new ArrayList();
+  }
+
+  public static class StopListenerFragment extends android.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      for (Runnable callback : callbacks) {
+        if (callback != null) {
+          callback.run();
+        }
+      }
+    }
+
+    void addCallback(Runnable callback) {
+      callbacks.add(callback);
+    }
+
+    private List<Runnable> callbacks = new ArrayList();
+  }
+
+  private static final String SUPPORT_FRAGMENT_TAG = "FirestoreOnStopObserverSupportFragment";
+  private static final String FRAGMENT_TAG = "FirestoreOnStopObserverFragment";
+
+  private void onActivityStopCallOnce(Activity activity, Runnable callback) {
+    // Android provides lifecycle callbacks (eg onStop()) that custom `Activity`s can extend. But we
```

Comment:
Most if not all of this description could move into the class comment for this type since it's a motivation for the design of the class as a whole.

Also on this line: it's not clear why "custom `Activity`s can extend" aren't usable to us. The issue (I believe) is that custom activities can add onStop behavior by subclassing, but we need to add onStop behavior to arbitrary activities with which we can't have a subclassing relationship.